pkg/boot/stboot/bootball.go

```diff
@@ -0,0 +1,574 @@
+	ball.signatures = make(map[string][]signature)
+	path := filepath.Join(ball.dir, signaturesDirName)
+
+	sigPool := make([]signature, 0)
+	err = filepath.Walk(path, func(path string, info os.FileInfo, err error) error {
+		ext := filepath.Ext(info.Name())
+
+		if !info.IsDir() && (ext == ".signature") {
+			sigBytes, err := ioutil.ReadFile(path)
+			if err != nil {
+				return err
+			}
+
+			dir := filepath.Dir(path)
+			index, err := strconv.Atoi(dir[len(dir)-1:])
+			if err != nil {
+				return err
+			}
+
+			certFile := strings.TrimSuffix(path, filepath.Ext(path)) + ".cert"
+			certBytes, err := ioutil.ReadFile(certFile)
+			if err != nil {
+				return err
+			}
+
+			cert, err := parseCertificate(certBytes)
+			if err != nil {
+				return err
+			}
+
+			sig := signature{
+				Bytes: sigBytes,
+				Cert:  cert,
+			}
+			sigPool = append(sigPool, sig)
+			key := ball.config.BootConfigs[index].Name
+			ball.signatures[key] = sigPool
+		}
+		return nil
+	})
+	if err != nil {
+		return
+	}
+	return
+}
+
+func validateFiles(prefix string, files ...string) (err error) {
+	for _, file := range files {
+		_, err = os.Stat(filepath.Join(prefix, file))
+		if err != nil {
+			return
+		}
+	}
+	return
+}
+
+func writeSignatures(sigs []signature, certFile, dir string) (err error) {
+	for i, sig := range sigs {
+		d := fmt.Sprintf("%s%d", bootFilesDirName, i)
+		path := filepath.Join(dir, signaturesDirName, d)
+		os.Mkdir(path, os.ModePerm)
+
+		id := fmt.Sprintf("%x", sig.Cert.PublicKey)[2:18]
+		sigName := fmt.Sprintf("%s.signature", id)
+		sigPath := filepath.Join(path, sigName)
+		werr := ioutil.WriteFile(sigPath, sig.Bytes, 0644)
+		if werr != nil {
+			return werr
+		}
+
+		certName := fmt.Sprintf("%s.cert", id)
+		certPath := filepath.Join(path, certName)
+		cerr := copyFile(certFile, certPath)
+		if cerr != nil {
+			return cerr
+		}
+	}
+	return
+}
+
+func makeConfigDir(cfg *Stconfig, origDir string) (dir string, err error) {
+	if err = validateFiles(cfg.RootCertPath); err != nil {
+		return
+	}
+
+	for _, bc := range cfg.BootConfigs {
+		if err = validateFiles(origDir, bc.FileNames()...); err != nil {
+			return
+		}
+	}
+
+	dir, err = ioutil.TempDir(os.TempDir(), "bootball")
+	if err != nil {
+		return
+	}
+
+	dstPath := filepath.Join(dir, signaturesDirName, rootCertName)
+	srcPath := filepath.Join(origDir, cfg.RootCertPath)
+	copyFile(srcPath, dstPath)
+
+	for i, bc := range cfg.BootConfigs {
+		dirName := fmt.Sprintf("%s%d", bootFilesDirName, i)
+		for _, file := range bc.FileNames() {
+			fileName := filepath.Base(file)
+			dstPath := filepath.Join(dir, dirName, fileName)
+			srcPath := filepath.Join(origDir, file)
+			copyFile(srcPath, dstPath)
+		}
+
+		bc.ChangeFilePaths(dirName)
+		cfg.BootConfigs[i] = bc
+	}
+
+	dstPath = filepath.Join(dir, ConfigName)
+	bytes, err := cfg.bytes()
+	if err != nil {
+		return
+	}
+	ioutil.WriteFile(dstPath, bytes, os.ModePerm)
+
+	return
+}
+
+func copyFile(src, dst string) (err error) {
```

Comment:
pkg/cp has a copy. cp.Copy(src, dst)

Reply:
It seems that cp.Copy does not create dst path